src/compiler/scala/tools/nsc/transform/TailCalls.scala

```diff
@@ -168,16 +165,37 @@ abstract class TailCalls extends Transform {
 
         label
       }
-      private def isRecursiveCall(t: Tree) = {
-        val receiver = t.symbol
-
-        (    (receiver != null)
-          && receiver.isMethod
-          && (method.name == receiver.name)
-          && (method.enclClass isSubClass receiver.enclClass)
-        )
+      // self-recursive calls, eagerly evaluated
+      object detectRecursion extends Traverser {
+        val detected = ListBuffer.empty[Tree]
+        override def traverse(tree: Tree) = tree match {
+          case Apply(fun, args) =>
+            if (isRecursiveCall(fun.symbol)) detected.addOne(tree)
+            traverse(fun)
+            for ((p, a) <- fun.symbol.paramLists.head.lazyZip(args))
+              traverse(a)
```

Comment:
```suggestion
            args.foreach(traverse)
```

?

Reply:
I saved the branch which previously said `if strict || !p.isByNameParam`...